revision.c

```diff
@@ -99,29 +99,125 @@ void mark_tree_uninteresting(struct repository *r, struct tree *tree)
 	mark_tree_contents_uninteresting(r, tree);
 }
 
+struct paths_and_oids {
+	struct string_list list;
```

Comment:
What is the expected size of the string list?  Is this O(number-of-directories-in-the-repo)?  Or just O(number-of-interesting-directories)?

Just wondering if this is very large, should we consider a hashmap.

But then, if you are processing them in sorted order the inserts would be appends??

Reply:
O(number of directories in a specific folder).

The inserts are not always appends, because we can have adds and deletes. I'll look into the hashmap aproach.

Reply:
Looks like this is O(number of paths within directories that have both uninteresting and interesting set).  I think.

Reply:
Wait, is your "struct paths_and_oids" a per-directory list of paths?  (I was thinking it was global.)

Reply:
it is per-directory. As we deepen, we recurse and create a new struct.

Reply:
and we fill it with _all subtrees we see_ and then use that list to determine if a path has both interesting and non-interesting in that set.